Valkyrie models

kirkkwang opened the pull request:
# Story

Ref:
- https://github.com/scientist-softserv/ams/issues/36

Generate valkyrie models of the following:
- Asset
- PhysicalInstantiation
- DigitalInstantiation
- EssenceTrack
- Contribution

Now we have:
- AssetResource
- PhysicalInstantiationResource
- DigitalInstantiationResource
- EssenceTrackResource
- ContributionResource

# Notes

Even after registering the test_adapter I was getting still getting errors so I'm skipping all the generated model test for now.
https://github.com/samvera/hyrax/blob/4.x-stable/spec/test_app_templates/lib/generators/test_app_generator.rb#L219-L228

Commit:
🎁 Generate new resources

This commit will run the follow commands:

```sh
rails g hyrax:work_resource AssetResource
rails g hyrax:work_resource PhysicalInstantiationResource
rails g hyrax:work_resource DigitalInstantiationResource
rails g hyrax:work_resource EssenceTrackResource
rails g hyrax:work_resource ContributionResource
```

This will generate new Valkyrie works for each of the existing work
types.

Note:  Because of dual booting I had to prepend `DEPENDENCIES_NEXT=1`

Commit:
🧹 Add attributes to valkyrie models

This commit will register the new models and also add the attributes.
The approach was to look at the fedora model and extract and conform the
attributes to the new valkyrie models as best as possible.  This is a
first pass and will most likely need to be updated as we go.

Commit:
🧹 Wrap dual boot conditional on generated files

This commit will add the dual boot conditional around the generated
files so that the build will not fail.

Adding script here for reference:

```rb
%w[
  app/controllers/hyrax/asset_resources_controller.rb
  app/controllers/hyrax/contribution_resources_controller.rb
  app/controllers/hyrax/digital_instantiation_resources_controller.rb
  app/controllers/hyrax/essence_track_resources_controller.rb
  app/controllers/hyrax/physical_instantiation_resources_controller.rb
  app/forms/asset_resource_form.rb
  app/forms/contribution_resource_form.rb
  app/forms/digital_instantiation_resource_form.rb
  app/forms/essence_track_resource_form.rb
  app/forms/physical_instantiation_resource_form.rb
  app/indexers/asset_resource_indexer.rb
  app/indexers/contribution_resource_indexer.rb
  app/indexers/digital_instantiation_resource_indexer.rb
  app/indexers/essence_track_resource_indexer.rb
  app/indexers/physical_instantiation_resource_indexer.rb
  app/models/asset_resource.rb
  app/models/contribution_resource.rb
  app/models/digital_instantiation_resource.rb
  app/models/essence_track_resource.rb
  app/models/physical_instantiation_resource.rb
  config/initializers/wings.rb
  spec/indexers/asset_resource_indexer_spec.rb
  spec/indexers/contribution_resource_indexer_spec.rb
  spec/indexers/digital_instantiation_resource_indexer_spec.rb
  spec/indexers/essence_track_resource_indexer_spec.rb
  spec/indexers/physical_instantiation_resource_indexer_spec.rb
  spec/models/asset_resource_spec.rb
  spec/models/contribution_resource_spec.rb
  spec/models/digital_instantiation_resource_spec.rb
  spec/models/essence_track_resource_spec.rb
  spec/models/physical_instantiation_resource_spec.rb
  spec/views/asset_resources/_asset_resource.html.erb_spec.rb
  spec/views/contribution_resources/_contribution_resource.html.erb_spec.rb
  spec/views/digital_instantiation_resources/_digital_instantiation_resource.html.erb_spec.rb
  spec/views/essence_track_resources/_essence_track_resource.html.erb_spec.rb
  spec/views/physical_instantiation_resources/_physical_instantiation_resource.html.erb_spec.rb
].each do |file|
  # open file
  FileUtils.cp file, "#{file}.bak"
  File.open("#{file}.bak", 'r') do |f|
    # read file
    lines = f.readlines
    # write file
    File.open(file, 'w') do |f|
      f.puts "# frozen_string_literal: true"
      f.puts "unless App.rails_5_1?"
      l = lines[1..-1] if lines[0].match "frozen"
      l ||= lines
      l.each do |line|
        f.puts "  " + line
      end
      f.puts "end"
    end
  end
end
```

Once files are looking good, delete the bak files:
```sh
find . -iname "*.bak" -delete -print
```

Commit:
✅ Skip newly generated model tests

This commit skips the newly generated model tests for now since we're
not fully in a position to test them yet.  I tried registering the
test_adapter as outlined in the generator but it led to other errors.

Ref:
  - https://github.com/samvera/hyrax/blob/4.x-stable/spec/test_app_templates/lib/generators/test_app_generator.rb#L219-L228

ShanaLMoore commented:
Should we upgrade Hyrax too or has that already been done?

kirkkwang commented:
That's gonna be tricky with our dual boot situation, Hyrax `main` depends on Ruby 3.2 or higher.

ShanaLMoore commented:
> That's gonna be tricky with our dual boot situation, Hyrax `main` depends on Ruby 3.2 or higher.

Oh right, we put a pause on that for a moment. Hmm 🤔 This looks fine for me but wonder if we should get Rob's eyes on it.

laritakr commented on config/metadata/asset_resource.yaml:
```diff
@@ -0,0 +1,357 @@
+      primary: false
+      multiple: true
+  intended_children_count:
+    type: string
```
This needs to include index_keys

kirkkwang replied:
additional indexing will be handled in another issue

Commit:
✅ Add rails_helper methods to support Valkyrie

This commit adds Valkyrie helpers taken from koppie which makes the
tests pass now.  Still keeping the App.rails_5_1? check for now.
